python/ray/serve/_private/application_state.py

```diff
@@ -51,15 +50,56 @@ def __init__(
             self.deployment_timestamp = time.time()
         self.deploy_obj_ref = deploy_obj_ref
         self.app_msg = ""
+        self.route_prefix = None
+
+        # As application keep updated, this is the variable to keep track
+        # old unused deployments deletion
+        self.deployments_to_delete = set()
 
     def delete(self):
         """Delete the application"""
         self.status = ApplicationStatus.DELETING
 
-    def deploy(self):
+    def deploy(self, deployment_params: List[Dict]) -> None:
         """Deploy the application"""
+
+        # Filter unused deployments from the new deployment_params
+        to_be_deployed_deployments = {params["name"] for params in deployment_params}
+        for deployment_name in self.get_all_deployments():
+            if deployment_name not in to_be_deployed_deployments:
+                self.deployment_state_manager.delete_deployment(deployment_name)
+                self.deployments_to_delete.add(deployment_name)
+        self.deployment_params = deployment_params
+
+        # Update route prefix for application
+        num_route_prefixes = 0
+        for deploy_param in deployment_params:
+            if (
+                "route_prefix" in deploy_param
+                and deploy_param["route_prefix"] is not None
+            ):
+                self.route_prefix = deploy_param["route_prefix"]
+                num_route_prefixes += 1
+        if num_route_prefixes > 1:
+            logger.warning(
+                f"Found multiple route prefix from application {self.name},"
+                " Other applications might overwrite route prefixes of "
```

Comment:
```suggestion
                "Other applications might overwrite these route prefixes. Please specify only one route prefix for the application to avoid this issue."
```